### PR TITLE
Guest OS Support: add JeOS/Fedora 22/23 [v4]

### PR DESCRIPTION
**Pull request by clebergnu**
This PR adds (late) support for Fedora 22 and 23, and also the JeOS guests based on it.

New JeOS images (based on Fedora 23), based on this configuration/kickstart have been pushed to the official asset repo.

--
Changes from v3 (#453):
   * Treat downloadable asset definition files more like config files
   * Simplified `data_dir.get_download_dir()` and introduced `data_dir.get_base_download_dir()`
   * Check for program/headers files earlier

Changes from v2 (#439):
   * Only overwrite asset files if newer than the user's version

Changes from v1:
   * Added development related packages to F22 and F23
   * Added Python 2 packages on F23
   * Set the JeOS.23 as the new default guest os
   * Removed the JeOS.20 asset (download) definition file
   * Fix the `data_dir.get_download_dir()` implementation and usage

### CODE REVIEW

File: shared/unattended/Fedora-22.ks [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ECHO "OS install is completed"
+grubby --remove-args="rhgb quiet" --update-kernel=$(grubby --default-kernel)
+dhclient
+chkconfig sshd on
+iptables -F
+systemctl mask tmp.mount
+echo 0 > /selinux/enforce
+sed -i "/^HWADDR/d" /etc/sysconfig/network-scripts/ifcfg-eth0
+dnf -y groupinstall c-development development-tools
+ECHO 'Post set up finished'
```
Comment (ldoktor) [inline shared/unattended/Fedora-22.ks]: ack

File: shared/unattended/JeOS-22.ks [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+sed -i "/^HWADDR/d" /etc/sysconfig/network-scripts/ifcfg-eth0
+sed -i -e "s,^GRUB_TIMEOUT=.*,GRUB_TIMEOUT=0," /etc/default/grub
+grub2-mkconfig > /etc/grub2.cfg
+yum install -y hdparm ntpdate qemu-guest-agent
+yum clean all
+mkdir -p /var/log/journal
+sed -i -e 's/\#SystemMaxUse=/SystemMaxUse=50M/g' /etc/systemd/journald.conf
+dd if=/dev/zero of=/fill-up-file bs=1M
+rm -f /fill-up-file
+ECHO 'Post set up finished'
```
Comment (ldoktor) [inline shared/unattended/JeOS-22.ks]: ack

File: shared/unattended/Fedora-23.ks [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ECHO "OS install is completed"
+grubby --remove-args="rhgb quiet" --update-kernel=$(grubby --default-kernel)
+dhclient
+chkconfig sshd on
+iptables -F
+systemctl mask tmp.mount
+echo 0 > /selinux/enforce
+sed -i "/^HWADDR/d" /etc/sysconfig/network-scripts/ifcfg-eth0
+dnf -y groupinstall c-development development-tools
+ECHO 'Post set up finished'
```
Comment (ldoktor) [inline shared/unattended/Fedora-23.ks]: ack

File: shared/unattended/JeOS-23.ks [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+sed -i "/^HWADDR/d" /etc/sysconfig/network-scripts/ifcfg-eth0
+sed -i -e "s,^GRUB_TIMEOUT=.*,GRUB_TIMEOUT=0," /etc/default/grub
+grub2-mkconfig > /etc/grub2.cfg
+yum install -y hdparm ntpdate qemu-guest-agent
+yum clean all
+mkdir -p /var/log/journal
+sed -i -e 's/\#SystemMaxUse=/SystemMaxUse=50M/g' /etc/systemd/journald.conf
+dd if=/dev/zero of=/fill-up-file bs=1M
+rm -f /fill-up-file
+ECHO 'Post set up finished'
```
Comment (ldoktor) [inline shared/unattended/JeOS-23.ks]: ack

File: virttest/defaults.py [hunk 5]
```diff
@@ -18,7 +18,7 @@ def get_default_guest_os_info():
     Gets the default asset and variant information
     TODO: Check for the ARCH and choose corresponding default asset
     """
-    return {'asset': 'jeos-21-64', 'variant': 'JeOS.21'}
+    return {'asset': 'jeos-23-64', 'variant': 'JeOS.23'}
```
Comment (ldoktor) [inline virttest/defaults.py]: I agree, out of curiosity, what JeOS images are there? (I dislike it does not support browsing) I noticed 21 and 23.

File: virttest/bootstrap.py [hunk 6]
```diff
@@ -753,6 +748,11 @@ def bootstrap(options, interactive=False):
 
     logging.info("")
     step += 1
+    logging.info("%d - Updating all test providers", step)
+    asset.download_all_test_providers(options.vt_update_providers)
+
+    logging.info("")
+    step += 1
```
Comment (ldoktor) [inline virttest/bootstrap.py]: makes sense

File: virttest/data_dir.py [hunk 7]
```diff
@@ -213,9 +213,11 @@ def get_tmp_dir(public=True):
     return tmp_dir
 
 
+def get_base_download_dir():
+    return BASE_DOWNLOAD_DIR
+
+
 def get_download_dir():
-    if not os.path.isdir(DOWNLOAD_DIR):
-        shutil.copytree(BASE_DOWNLOAD_DIR, DOWNLOAD_DIR)
```
Comment (ldoktor) [inline virttest/data_dir.py]: looks good

Comment (ldoktor) [inline virttest/data_dir.py]: (hopefully it won't break custom scripts...)

File: virttest/bootstrap.py [hunk 8]
```diff
@@ -262,6 +262,40 @@ def get_directory_structure(rootdir, guest_file):
...
+    logging.debug("Copying downloadable assets file definitions from %s "
+                  "into %s", base_download_dir, download_dir)
+    download_file_list = glob.glob(os.path.join(base_download_dir,
+                                                "*.ini"))
+    logging.debug("Donwload file list: %s", download_file_list)
+    for src_file in download_file_list:
+        dst_file = os.path.join(download_dir,
+                                os.path.basename(src_file))
+        if not os.path.isfile(dst_file):
+            shutil.copyfile(src_file, dst_file)
```
Comment (ldoktor) [inline virttest/bootstrap.py]: create_config_file was more verbose here, but I don't see the urge to log this...

File: virttest/bootstrap.py [hunk 9]
```diff
@@ -262,6 +262,40 @@ def get_directory_structure(rootdir, guest_file):
...
+        else:
+            diff_cmd = "diff -Naur %s %s" % (dst_file, src_file)
+            diff_result = process.run(
+                diff_cmd, ignore_status=True, verbose=False)
+            if diff_result.exit_status != 0:
+                logging.info("%s result:\n %s",
+                             diff_result.command, diff_result.stdout)
+                answer = genio.ask('Download file "%s" differs from "%s". '
+                                   'Overwrite?' % (dst_file, src_file),
+                                   auto=not interactive)
```
Comment (ldoktor) [inline virttest/bootstrap.py]: If we deal with them as config files, shouldn't the `--vt-update-config` apply here?

File: virttest/bootstrap.py [hunk 10]
```diff
@@ -262,6 +262,40 @@ def get_directory_structure(rootdir, guest_file):
...
+                if answer == "y":
+                    logging.debug("Restoring download file %s from sample",
+                                  dst_file)
+                    shutil.copyfile(src_file, dst_file)
+                else:
+                    logging.debug("Preserving existing %s file", dst_file)
+            else:
+                logging.debug('Download file %s exists, not touching',
```
Comment (ldoktor) [inline virttest/bootstrap.py]: looks similar to `create_config_files`, let me check if that works :smile:

Comment (ldoktor) [inline virttest/bootstrap.py]: Yep it works fine. Please let me know what you think about the `--vt-update-config`. I think it should apply here too, but it's not strong preference. Apart from this I'm ready to merge this PR.

File: virttest/defaults.py (hunk 5 above)
Comment (clebergnu) [inline virttest/defaults.py]: I also dislike it, but as I said before, it's a current hosting limitation. I'm working on a better hosting though.

File: virttest/bootstrap.py (hunk 9 above)
Comment (clebergnu) [inline virttest/bootstrap.py]: I said that we're treating them "more like config files", but I'm not really considering them config files, as Avocado-VT users already have a fixed understanding of what config (.cfg) files are.  I wouldn't add yet another command line option for this, `--yes-to-all` already works fine here.

Comment (ldoktor) [inline virttest/bootstrap.py]: OK

**Pull request merged**